### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by jtopjian**
As I mentioned in the Terraform repository, I began refactoring the cobbler client package in order to more easily model Terraform resources. It was a much larger endeavor than I imagined, but was actually quite fun. :smile: 

I'm not finished -- I just wanted to publish some code to show you guys what I've been doing.

At the moment, you should be able to use this package to Add and Get systems from Cobbler. You can use the `cmd/main.go` command as an example. 

The following work still needs done to "systems":
1. More unit tests for systems.
2. Delete functionality for systems.

After that, I plan to work on Distros and Profiles, though that shouldn't take long at all. The bulk of the learning curve is over.

Let me know what you guys think!

### CODE REVIEW

**Comment by thijsschnitger** [comment]
Thanks, that's quite a restructure. We contemplated switching the xpath stuff out for an xmlrpc lib, nice that you did that.
I won't merge just yet because it currently breaks the tests on snippets and kickstartfiles, we'll wait for the rest of your stuff.

**Comment by jtopjian** [comment]
No problem at all. I'm sorry for making so many changes -- they seemed to all flow from one to the other.

I've been traveling for work (I'll be in Utrecht tomorrow, actually :smile: ), so I haven't had a lot of time to keep working on this, but I do have plans to finish it up.

**Comment by frankscholten** [comment]
@jtopjian Utrecht? Cool my hometown :-) Btw feel free to drop by the CS office in Amsterdam.

**Comment by jtopjian** [comment]
I've made some more progress!
- Snippet and Kickstart CRUD
- Distro CRUD
- System Delete
- Typed interface support

`cmd/main.go` can act as an acceptance test suite for the unit tests that are unable to fully test workflows.

At this point, I'm going to start modeling Terraform resources and see how far I get. That should be a good "production" test of this client library. It's very possible that changes will be made in order to get the modeling to work.

File: cobblerclient.go [hunk 1]
```diff
@@ -74,101 +102,101 @@ func (c *Client) Login() (bool, error) {
...
 	}
+}
 
-	defer res.Body.Close()
-	body, err := ioutil.ReadAll(res.Body)
-	if err != nil {
-		return nil, err
-	}
+// cobblerDataHacks is a hook for the mapstructure decoder.
+// It's used to smooth out issues with converting fields and types from Cobbler.
```
Comment (mongrelion) [inline cobblerclient.go:120]: Can you please add a short comment/snippet on how to use this function?

File: fixtures/get-kickstart-file-res.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version='1.0'?>
```
Comment (mongrelion) [inline fixtures/get-kickstart-file-res.xml:1]: Maybe indent this file to keep things uniform :+1:

File: fixtures/login-res.xml [hunk 3]
```diff
@@ -1,10 +1 @@
-<?xml version='1.0'?>
-<methodResponse>
-  <params>
-    <param>
-      <value>
-        <string>sa/1EWr40BWU+Pq3VEOOpD4cQtxkeMuFUw==</string>
-      </value>
-    </param>
-  </params>
-</methodResponse>
```
Comment (mongrelion) [inline fixtures/login-res.xml:10]: What's the reason for this change? Before it was more readable. Now it's a really long line and makes it more difficult to follow.

File: fixtures/sync-res.xml [hunk 4]
```diff
@@ -1,8 +1 @@
-<?xml version='1.0'?>
-<methodResponse>
-  <params>
-    <param>
-      <value><boolean>1</boolean></value>
-    </param>
-  </params>
-</methodResponse>
```
Comment (mongrelion) [inline fixtures/sync-res.xml:8]: Same for this one. I don't understand why we're moving from a readable approach to oneliners.

**Comment by jtopjian** [comment]
@mongrelion The latest commit addresses your comments.

Regarding the two files that were flattened to a single line, this was in response to the xmlrpc library not flattening files internally. I ended up taking care of this globally [here](https://github.com/jtopjian/cobblerclient/blob/refactor/cobblerclient_test.go#L39-L42) but must not have reverted all files.

I'm happy to address any other comments or concerns.

**Comment by mongrelion** [comment]
Hello, @jtopjian. Thanks for this pull request. We appreciate it.
It's quite a long list of changes. I somehow would have preferred if you submitted smaller changes/PRs but I don't want to discourage you or undervalue your contribution. What I'm trying to say is that it's going to take me a little to test it all out as I currently am busy with other projects but I'm doing my best to make some time for this changes to be merged.

Next step for me is to test this out in my local machine, make sure that everything is working and come up with a release plan and coordinate with you so that we don't break the Cobbler provider for Terraform.

**Comment by jtopjian** [comment]
@mongrelion That's fine. I'm going to proceed with working on the Cobbler Terraform provider to be included in the main Terraform codebase. I can work out of my github.com/jtopjian namespace for the time being.

File: cmd/main.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+
+	cobbler "github.com/ContainerSolutions/cobblerclient"
+)
+
+var config = cobbler.ClientConfig{
```
Comment (mongrelion) [inline cmd/main.go:10]: Maybe it's a good idea to use the `flag` package here to make all these variables overridable. Can't test in my local env. without modifying the source.

File: cmd/main.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+		Initrd:    "/var/www/cobbler/ks_mirror/Ubuntu-14.04/install/netboot/ubuntu-installer/amd64/initrd.gz",
+	}
+
+	fmt.Println("Creating a Distro")
+	newDistro, err := c.CreateDistro(d)
+	if err != nil {
+		fmt.Printf("Error creating distro: %s\n", err)
+	}
+
+	fmt.Printf("New Distro: %+v\n", newDistro)
```
Comment (mongrelion) [inline cmd/main.go:40]: This line and the rest below it that reference to `newDistro` will fail if the distro is not created.

**Comment by mongrelion** [comment]
I'm getting the following error while creating a distro:
```
Token: UxgxXi7sPFCefKoHPEdY+huu2YLU8qUUmQ==
Creating a Distro
Error creating distro: error: "<class 'cobbler.cexceptions.CX'>:'kernel not found: /var/www/cobbler/ks_mirror/Ubuntu-14.04/install/netboot/ubuntu-installer/amd64/linux'" code: 1
```

Could you perhaps provide a way for feeding the test data into Cobbler prior to running the `cmd/main.go` file?

File: cmd/main.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	newSystem, err := c.CreateSystem(s)
+	if err != nil {
+		fmt.Printf("Error creating system: %s\n", err)
+	}
+
+	fmt.Printf("New System: %+v\n", newSystem)
+
+	fmt.Println("Adding NIC to System")
```
Comment (mongrelion) [inline cmd/main.go:92]: This line and the rest below it that reference to `newSystem` will fail if the system is not created.

**Comment by mongrelion** [comment]
A couple of things: 
Creating and deleting a snippet works
Creating and deleting a kickstart file works

However, while creating a distro I'm getting the following error:

```
Token: UxgxXi7sPFCefKoHPEdY+huu2YLU8qUUmQ==
Creating a Distro
Error creating distro: error: "<class 'cobbler.cexceptions.CX'>:'kernel not found: /var/www/cobbler/ks_mirror/Ubuntu-14.04/install/netboot/ubuntu-installer/amd64/linux'" code: 1
```

I also get the following error in the lines where the file tries to create a system:

```
Token: XEnQMYa1fPJANG7jIfbzw/qrp5TKtA5adg==
Creating a System
Error creating system: error: "<class 'cobbler.cexceptions.CX'>:'invalid profile name: Ubuntu-14.04-x86_64'" code: 1
```

Could you perhaps provide a way for feeding the test data into Cobbler prior to running the `cmd/main.go` file?

**Comment by jtopjian** [comment]
Yes, `cmd/main.go` is really just a scratchpad to test various functionality that is a step beyond standard unit testing. I never gave much thought to making it customizable for all sorts of environments -- I assumed that it would be easy for someone to change themselves. It's the same for early failures, too: early failures will not stop other tests from running. 

tl;dr: `cmd/main.go` is far from a full acceptance test suite. :smile: 

As for the reason why those areas are failing for you: `cobbler import` was already run to import a basic distribution, so it's assuming that there are distro files under `/var/www` and such.

With all of that in mind, this can be used to generate a replica of the environment that I am testing with: https://github.com/jtopjian/terraform-cobbler-acceptance. Once that environment is set up, `cmd/main.go` will successfully run.
